shared/src/main/java/de/pnp/manager/network/state/StateMachine.java

```diff
@@ -0,0 +1,73 @@
+package de.pnp.manager.network.state;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class StateMachine<Event> {
+
+    protected Set<State> states;
+    protected State currentState;
+    protected Collection<Transition<Event>> transitions;
+    protected Consumer<Event> onNoTransition;
+
+    public StateMachine(Set<State> states, State start) {
+        this.states = states;
+        this.currentState = start;
+        this.transitions = new ArrayList<>();
+        this.onNoTransition = null;
+    }
+
+    public boolean fire(Event event) {
+
+        for (Transition<Event> transition : transitions) {
+            if (
+                    transition.from == currentState &&
+                            transition.eventCheck.apply(event) &&
+                            states.contains(transition.to)
+            ) {
+                if (transition.eventHandler != null) {
+                    transition.eventHandler.accept(event);
+                }
+                this.currentState = transition.to;
+                return true;
+            }
+        }
```

Comment:
would be faster to have a map 
state -> transition
then we only need to check a subset of transitions